File: src/backtest/engine.ts
import { KiteHistoricalProvider } from "../market_data/kite_historical_provider.js";
import { SignalBuilder } from "../signal/signal_builder.js";
import { MarketBar, ScreenerResult } from "../types.js";
import { atr, ema, pctChange, rsi, sma } from "../utils/indicators.js";

type OpenTrade = {
  symbol: string;
  entryTime: string;
  entryPrice: number;
  entryStop: number;
  targetPrice: number;
  qty: number;
  daysHeld: number;
};

type ClosedTrade = {
  symbol: string;
  entryTime: string;
  exitTime: string;
  entryPrice: number;
  exitPrice: number;
  qty: number;
  pnl: number;
  rMultiple: number;
  reason: string;
};

export type BacktestConfig = {
  from: string;
  to: string;
  symbols: string[];
  initialCapital: number;
  maxHoldDays: number;
  slippageBps: number;
  feeBps: number;
  maxOpenPositions: number;
  minRsi: number;
  breakoutBufferPct: number;
  atrStopMultiple: number;
  riskPerTrade: number;
  minAdv20: number;
  minVolumeRatio: number;
  maxSignals: number;
};

export type BacktestRunResult = {
  runId: string;
  startedAt: string;
  finishedAt: string;
  from: string;
  to: string;
  symbols: string[];
  initialCapital: number;
  finalCapital: number;
  totalPnl: number;
  trades: number;
  winRate: number;
  avgR: number;
  expectancy: number;
  maxDrawdownAbs: number;
  maxDrawdownPct: number;
  cagrPct: number;
  sharpeProxy: number;
  bySymbol: Array<{
    symbol: string;
    trades: number;
    winRate: number;
    pnl: number;
    avgR: number;
  }>;
  equityCurve: Array<{
    date: string;
    equity: number;
    realizedPnl: number;
    unrealizedPnl: number;
    openPositions: number;
  }>;
  notes: string[];
};

const DEFAULT_SYMBOLS = [
  "RELIANCE",
  "TCS",
  "INFY",
  "HDFCBANK",
  "ICICIBANK",
  "SBIN",
  "LT",
  "AXISBANK",
  "KOTAKBANK",
  "ITC",
  "TATAMOTORS",
  "TATASTEEL"
];

export async function runBacktest(
  provider: KiteHistoricalProvider,
  cfg: Partial<BacktestConfig>
): Promise<BacktestRunResult> {
  const startedAt = new Date().toISOString();
  const runId = `bt-${Date.now()}`;
  const config = resolveConfig(cfg);

  const barsBySymbol = await loadBars(provider, config.symbols, config.from, config.to);
  const usableSymbols = Array.from(barsBySymbol.keys());
  const notes: string[] = [];
  if (usableSymbols.length === 0) {
    throw new Error("No symbols have enough historical bars for selected range");
  }
  if (usableSymbols.length < config.symbols.length) {
    notes.push(
      `Skipped ${config.symbols.length - usableSymbols.length} symbols due to missing/short history.`
    );
  }

  const tradingDays = buildTradingCalendar(barsBySymbol);
  const signalBuilder = new SignalBuilder({
    minRsi: config.minRsi,
    breakoutBufferPct: config.breakoutBufferPct,
    atrStopMultiple: config.atrStopMultiple,
    riskPerTrade: config.riskPerTrade,
    minAdv20: config.minAdv20,
    minVolumeRatio: config.minVolumeRatio,
    maxSignals: config.maxSignals
  });

  let realizedPnl = 0;
  const openTrades = new Map<string, OpenTrade>();
  const closedTrades: ClosedTrade[] = [];
  const equityCurve: BacktestRunResult["equityCurve"] = [];

  for (const day of tradingDays) {
    const bySymbol = buildDayBarMap(barsBySymbol, day);

    // Exit checks run before fresh entries to avoid same-day re-entry churn.
    for (const trade of Array.from(openTrades.values())) {
      const bar = bySymbol.get(trade.symbol);
      if (!bar) {
        trade.daysHeld += 1;
        continue;
      }
      trade.daysHeld += 1;
      const exit = resolveExit(bar, trade, config.maxHoldDays);
      if (!exit) {
        continue;
      }

      const exitPrice = applySlippage(exit.price, "SELL", config.slippageBps);
      const turnover = trade.entryPrice * trade.qty + exitPrice * trade.qty;
      const fees = turnover * (config.feeBps / 10_000);
      const pnl = (exitPrice - trade.entryPrice) * trade.qty - fees;
      realizedPnl += pnl;
      const riskPerShare = Math.max(trade.entryPrice - trade.entryStop, 0.01);
      const rMultiple = (exitPrice - trade.entryPrice) / riskPerShare;

      closedTrades.push({
        symbol: trade.symbol,
        entryTime: trade.entryTime,
        exitTime: bar.time,
        entryPrice: trade.entryPrice,
        exitPrice,
        qty: trade.qty,
        pnl,
        rMultiple,
        reason: exit.reason
      });
      openTrades.delete(trade.symbol);
    }

    const currentEquity = config.initialCapital + realizedPnl;
    const candidates = buildCandidatesForDay(barsBySymbol, day);
    const signals = signalBuilder.buildSignals(candidates, currentEquity);
    for (const signal of signals) {
      if (openTrades.size >= config.maxOpenPositions) {
        break;
      }
      if (openTrades.has(signal.symbol) || signal.qty <= 0) {
        continue;
      }
      const bar = bySymbol.get(signal.symbol);
      if (!bar) {
        continue;
      }
      const entryPrice = applySlippage(signal.entryPrice, "BUY", config.slippageBps);
      openTrades.set(signal.symbol, {
        symbol: signal.symbol,
        entryTime: bar.time,
        entryPrice,
        entryStop: signal.stopPrice,
        targetPrice: signal.targetPrice ?? signal.entryPrice * 1.04,
        qty: signal.qty,
        daysHeld: 0
      });
    }

    const unrealizedPnl = computeUnrealizedPnl(openTrades, bySymbol);
    const equity = config.initialCapital + realizedPnl + unrealizedPnl;
    equityCurve.push({
      date: day,
      equity,
      realizedPnl,
      unrealizedPnl,
      openPositions: openTrades.size
    });
  }

  const finalMap = buildDayBarMap(barsBySymbol, tradingDays[tradingDays.length - 1] ?? config.to);
  for (const trade of Array.from(openTrades.values())) {
    const bar = finalMap.get(trade.symbol);
    if (!bar) {
      continue;
    }
    const exitPrice = applySlippage(bar.close, "SELL", config.slippageBps);
    const turnover = trade.entryPrice * trade.qty + exitPrice * trade.qty;
    const fees = turnover * (config.feeBps / 10_000);
    const pnl = (exitPrice - trade.entryPrice) * trade.qty - fees;
    realizedPnl += pnl;
    const riskPerShare = Math.max(trade.entryPrice - trade.entryStop, 0.01);
    const rMultiple = (exitPrice - trade.entryPrice) / riskPerShare;
    closedTrades.push({
      symbol: trade.symbol,
      entryTime: trade.entryTime,
      exitTime: bar.time,
      entryPrice: trade.entryPrice,
      exitPrice,
      qty: trade.qty,
      pnl,
      rMultiple,
      reason: "forced_eod"
    });
    openTrades.delete(trade.symbol);
  }

  const finalCapital = config.initialCapital + realizedPnl;
  const metrics = computeMetrics(config.initialCapital, finalCapital, closedTrades, equityCurve);
  const finishedAt = new Date().toISOString();

  return {
    runId,
    startedAt,
    finishedAt,
    from: config.from,
    to: config.to,
    symbols: usableSymbols,
    initialCapital: config.initialCapital,
    finalCapital,
    totalPnl: realizedPnl,
    trades: closedTrades.length,
    winRate: metrics.winRate,
    avgR: metrics.avgR,
    expectancy: metrics.expectancy,
    maxDrawdownAbs: metrics.maxDrawdownAbs,
    maxDrawdownPct: metrics.maxDrawdownPct,
    cagrPct: metrics.cagrPct,
    sharpeProxy: metrics.sharpeProxy,
    bySymbol: summarizeBySymbol(closedTrades),
    equityCurve,
    notes
  };
}

async function loadBars(
  provider: KiteHistoricalProvider,
  symbols: string[],
  from: string,
  to: string
): Promise<Map<string, MarketBar[]>> {
  const instruments = await provider.getInstruments("NSE");
  const out = new Map<string, MarketBar[]>();
  for (const symbol of symbols) {
    const inst = instruments.get(symbol);
    if (!inst) {
      continue;
    }
    try {
      const bars = await provider.getHistoricalDayBars(inst.instrumentToken, from, to);
      const normalized = bars
        .map((b) => ({ ...b, symbol }))
        .filter((b) => Number.isFinite(b.close) && Number.isFinite(b.volume))
        .sort((a, b) => a.time.localeCompare(b.time));
      if (normalized.length >= 80) {
        out.set(symbol, normalized);
      }
    } catch {
      // Skip symbols that fail due to temporary API/instrument issues.
    }
  }
  return out;
}

function buildTradingCalendar(barsBySymbol: Map<string, MarketBar[]>) {
  const dates = new Set<string>();
  for (const bars of barsBySymbol.values()) {
    for (const bar of bars) {
      dates.add(bar.time.slice(0, 10));
    }
  }
  return Array.from(dates).sort();
}

function buildDayBarMap(barsBySymbol: Map<string, MarketBar[]>, day: string) {
  const out = new Map<string, MarketBar>();
  for (const [symbol, bars] of barsBySymbol.entries()) {
    const bar = bars.find((x) => x.time.slice(0, 10) === day);
    if (bar) {
      out.set(symbol, bar);
    }
  }
  return out;
}

function buildCandidatesForDay(
  barsBySymbol: Map<string, MarketBar[]>,
  day: string
): ScreenerResult[] {
  const out: ScreenerResult[] = [];
  for (const [symbol, bars] of barsBySymbol.entries()) {
    const idx = bars.findIndex((b) => b.time.slice(0, 10) === day);
    if (idx < 61) {
      continue;
    }
    const slice = bars.slice(0, idx + 1);
    if (slice.length < 80) {
      continue;
    }
    const closes = slice.map((b) => b.close);
    const volumes = slice.map((b) => b.volume);
    const highs = slice.map((b) => b.high);
    const lows = slice.map((b) => b.low);
    try {
      const close = closes[closes.length - 1];
      const ema20 = ema(closes, 20);
      const ema50 = ema(closes, 50);
      const rsi14 = rsi(closes, 14);
      const atr14 = atr(highs, lows, closes, 14);
      const high20 = Math.max(...highs.slice(-20));
      const adv20 = averageTradedValue(closes.slice(-20), volumes.slice(-20));
      const volumeRatio = sma(volumes.slice(-20)) / sma(volumes.slice(-50));
      const rsScore60d = pctChange(closes[closes.length - 61], close);
      out.push({
        symbol,
        close,
        ema20,
        ema50,
        rsi14,
        atr14,
        rsScore60d,
        adv20,
        volumeRatio,
        high20
      });
    } catch {
      // Ignore symbol on malformed series.
    }
  }
  return out.sort((a, b) => b.rsScore60d - a.rsScore60d);
}

function averageTradedValue(closes: number[], volumes: number[]) {
  return sma(closes.map((close, i) => close * volumes[i]));
}

function resolveExit(
  bar: MarketBar,
  trade: OpenTrade,
  maxHoldDays: number
): { price: number; reason: string } | null {
  if (bar.low <= trade.entryStop) {
    return { price: trade.entryStop, reason: "stop_loss" };
  }
  if (bar.high >= trade.targetPrice) {
    return { price: trade.targetPrice, reason: "target_hit" };
  }
  if (trade.daysHeld >= maxHoldDays) {
    return { price: bar.close, reason: "max_hold" };
  }
  return null;
}

function applySlippage(price: number, side: "BUY" | "SELL", slippageBps: number) {
  const factor = slippageBps / 10_000;
  return side === "BUY" ? price * (1 + factor) : price * (1 - factor);
}

function computeUnrealizedPnl(openTrades: Map<string, OpenTrade>, dayBars: Map<string, MarketBar>) {
  let total = 0;
  for (const trade of openTrades.values()) {
    const bar = dayBars.get(trade.symbol);
    if (!bar) {
      continue;
    }
    total += (bar.close - trade.entryPrice) * trade.qty;
  }
  return total;
}

function computeMetrics(
  initialCapital: number,
  finalCapital: number,
  trades: ClosedTrade[],
  equityCurve: BacktestRunResult["equityCurve"]
) {
  const wins = trades.filter((t) => t.pnl > 0).length;
  const winRate = trades.length === 0 ? 0 : wins / trades.length;
  const avgR = avg(trades.map((t) => t.rMultiple));
  const expectancy = avg(trades.map((t) => t.pnl));
  const maxDrawdownAbs = maxDrawdownAbsFromEquity(equityCurve.map((x) => x.equity));
  const maxDrawdownPct = maxDrawdownPctFromEquity(equityCurve.map((x) => x.equity));
  const cagrPct = calcCagr(equityCurve, initialCapital, finalCapital);
  const sharpeProxy = calcSharpeProxy(equityCurve);
  return { winRate, avgR, expectancy, maxDrawdownAbs, maxDrawdownPct, cagrPct, sharpeProxy };
}

function summarizeBySymbol(trades: ClosedTrade[]) {
  const map = new Map<string, { trades: number; wins: number; pnl: number; r: number }>();
  for (const trade of trades) {
    const current = map.get(trade.symbol) ?? { trades: 0, wins: 0, pnl: 0, r: 0 };
    current.trades += 1;
    if (trade.pnl > 0) {
      current.wins += 1;
    }
    current.pnl += trade.pnl;
    current.r += trade.rMultiple;
    map.set(trade.symbol, current);
  }
  return Array.from(map.entries())
    .map(([symbol, x]) => ({
      symbol,
      trades: x.trades,
      winRate: x.trades === 0 ? 0 : x.wins / x.trades,
      pnl: x.pnl,
      avgR: x.trades === 0 ? 0 : x.r / x.trades
    }))
    .sort((a, b) => b.pnl - a.pnl);
}

function resolveConfig(cfg: Partial<BacktestConfig>): BacktestConfig {
  return {
    from: cfg.from ?? process.env.BACKTEST_FROM ?? dateOffset(365),
    to: cfg.to ?? process.env.BACKTEST_TO ?? todayDate(),
    symbols: cfg.symbols ?? parseCsv(process.env.BACKTEST_SYMBOLS, DEFAULT_SYMBOLS),
    initialCapital: cfg.initialCapital ?? Number(process.env.BACKTEST_INITIAL_CAPITAL ?? "1000000"),
    maxHoldDays: cfg.maxHoldDays ?? Number(process.env.BACKTEST_MAX_HOLD_DAYS ?? "15"),
    slippageBps: cfg.slippageBps ?? Number(process.env.BACKTEST_SLIPPAGE_BPS ?? "5"),
    feeBps: cfg.feeBps ?? Number(process.env.BACKTEST_FEE_BPS ?? "12"),
    maxOpenPositions: cfg.maxOpenPositions ?? Number(process.env.BACKTEST_MAX_OPEN_POSITIONS ?? "5"),
    minRsi: cfg.minRsi ?? Number(process.env.STRATEGY_MIN_RSI ?? "55"),
    breakoutBufferPct: cfg.breakoutBufferPct ?? Number(process.env.STRATEGY_BREAKOUT_BUFFER_PCT ?? "0.02"),
    atrStopMultiple: cfg.atrStopMultiple ?? Number(process.env.ATR_STOP_MULTIPLE ?? "2"),
    riskPerTrade: cfg.riskPerTrade ?? Number(process.env.RISK_PER_TRADE ?? "0.015"),
    minAdv20: cfg.minAdv20 ?? Number(process.env.STRATEGY_MIN_ADV20 ?? "100000000"),
    minVolumeRatio: cfg.minVolumeRatio ?? Number(process.env.STRATEGY_MIN_VOLUME_RATIO ?? "1.2"),
    maxSignals: cfg.maxSignals ?? Number(process.env.STRATEGY_MAX_SIGNALS ?? "5")
  };
}

function parseCsv(value: string | undefined, fallback: string[]) {
  if (!value || value.trim().length === 0) {
    return fallback;
  }
  return value
    .split(",")
    .map((x) => x.trim().toUpperCase())
    .filter((x) => x.length > 0);
}

function maxDrawdownAbsFromEquity(series: number[]) {
  let peak = Number.NEGATIVE_INFINITY;
  let maxDd = 0;
  for (const v of series) {
    peak = Math.max(peak, v);
    maxDd = Math.max(maxDd, peak - v);
  }
  return maxDd;
}

function maxDrawdownPctFromEquity(series: number[]) {
  let peak = Number.NEGATIVE_INFINITY;
  let maxDd = 0;
  for (const v of series) {
    peak = Math.max(peak, v);
    if (peak > 0) {
      maxDd = Math.max(maxDd, (peak - v) / peak);
    }
  }
  return maxDd;
}

function calcCagr(
  equityCurve: BacktestRunResult["equityCurve"],
  initialCapital: number,
  finalCapital: number
) {
  if (equityCurve.length < 2 || initialCapital <= 0 || finalCapital <= 0) {
    return 0;
  }
  const start = new Date(`${equityCurve[0].date}T00:00:00.000Z`).getTime();
  const end = new Date(`${equityCurve[equityCurve.length - 1].date}T00:00:00.000Z`).getTime();
  const years = Math.max((end - start) / (365.25 * 24 * 3600 * 1000), 1 / 252);
  return (Math.pow(finalCapital / initialCapital, 1 / years) - 1) * 100;
}

function calcSharpeProxy(equityCurve: BacktestRunResult["equityCurve"]) {
  if (equityCurve.length < 3) {
    return 0;
  }
  const returns: number[] = [];
  for (let i = 1; i < equityCurve.length; i += 1) {
    const prev = equityCurve[i - 1].equity;
    const curr = equityCurve[i].equity;
    if (prev > 0) {
      returns.push((curr - prev) / prev);
    }
  }
  const mu = avg(returns);
  const sigma = std(returns);
  if (sigma === 0) {
    return 0;
  }
  return (mu / sigma) * Math.sqrt(252);
}

function avg(values: number[]) {
  if (values.length === 0) {
    return 0;
  }
  return values.reduce((acc, v) => acc + v, 0) / values.length;
}

function std(values: number[]) {
  if (values.length < 2) {
    return 0;
  }
  const mu = avg(values);
  const variance = avg(values.map((v) => (v - mu) ** 2));
  return Math.sqrt(variance);
}

function todayDate() {
  const d = new Date();
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(
    d.getDate()
  ).padStart(2, "0")}`;
}

function dateOffset(days: number) {
  const d = new Date();
  d.setDate(d.getDate() - days);
  return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(
    d.getDate()
  ).padStart(2, "0")}`;
}

File: python_app/algo_trading_py/backtest/engine.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

from algo_trading_py.market_data.kite_historical_provider import KiteHistoricalProvider
from algo_trading_py.utils.indicators import atr, ema, pct_change, rsi, sma, std


@dataclass(slots=True)
class BacktestConfig:
  from_date: str
  to_date: str
  symbols: list[str]
  initial_capital: float = 1_000_000
  max_hold_days: int = 15
  slippage_bps: float = 5
  fee_bps: float = 12
  max_open_positions: int = 5
  min_rsi: float = 55
  breakout_buffer_pct: float = 0.02
  atr_stop_multiple: float = 2
  risk_per_trade: float = 0.015
  max_results: int = 5


@dataclass(slots=True)
class BacktestResult:
  from_date: str
  to_date: str
  symbols: list[str]
  trades: int
  win_rate: float
  total_pnl: float
  max_drawdown_abs: float
  max_drawdown_pct: float
  cagr_pct: float
  sharpe_proxy: float


def run_backtest(provider: KiteHistoricalProvider, cfg: BacktestConfig) -> BacktestResult:
  history_from = _offset_from(cfg.from_date, 180)
  bars_by_symbol = _load_bars(provider, cfg.symbols, history_from, cfg.from_date, cfg.to_date)
  symbols = [s for s, bars in bars_by_symbol.items() if bars]
  if not symbols:
    raise RuntimeError("No symbols have enough historical bars for selected range")
  trading_days = sorted({b["time"][:10] for bars in bars_by_symbol.values() for b in bars if cfg.from_date <= b["time"][:10] <= cfg.to_date})
  open_pos: dict[str, dict[str, float | int | str]] = {}
  equity_curve: list[float] = []
  realized = 0.0
  pnls: list[float] = []
  for day in trading_days:
    day_map = {s: next((x for x in bars if x["time"][:10] == day), None) for s, bars in bars_by_symbol.items()}
    # exits
    for s in list(open_pos.keys()):
      pos = open_pos[s]
      bar = day_map.get(s)
      if bar is None:
        pos["days_held"] = int(pos["days_held"]) + 1
        continue
      pos["days_held"] = int(pos["days_held"]) + 1
      exit_price = None
      if bar["low"] <= float(pos["stop"]):
        exit_price = float(pos["stop"])
      elif bar["high"] >= float(pos["target"]):
        exit_price = float(pos["target"])
      elif int(pos["days_held"]) >= cfg.max_hold_days:
        exit_price = float(bar["close"])
      if exit_price is None:
        continue
      filled = _slippage(exit_price, "SELL", cfg.slippage_bps)
      qty = float(pos["qty"])
      entry = float(pos["entry"])
      turnover = (entry * qty) + (filled * qty)
      fees = turnover * (cfg.fee_bps / 10_000)
      pnl = (filled - entry) * qty - fees
      realized += pnl
      pnls.append(pnl)
      del open_pos[s]
    # entries
    if len(open_pos) < cfg.max_open_positions:
      candidates = _build_candidates(day, bars_by_symbol, cfg)
      for c in candidates[: cfg.max_results]:
        if c["symbol"] in open_pos or len(open_pos) >= cfg.max_open_positions:
          continue
        bar = day_map.get(c["symbol"])
        if bar is None:
          continue
        entry = _slippage(c["entry"], "BUY", cfg.slippage_bps)
        open_pos[c["symbol"]] = {
          "entry": entry,
          "stop": c["stop"],
          "target": c["target"],
          "qty": c["qty"],
          "days_held": 0,
        }
    unrealized = 0.0
    for s, pos in open_pos.items():
      bar = day_map.get(s)
      if bar:
        unrealized += (float(bar["close"]) - float(pos["entry"])) * float(pos["qty"])
    equity_curve.append(cfg.initial_capital + realized + unrealized)
  final_cap = cfg.initial_capital + realized
  returns = []
  for i in range(1, len(equity_curve)):
    p = equity_curve[i - 1]
    c = equity_curve[i]
    if p > 0:
      returns.append((c - p) / p)
  return BacktestResult(
    from_date=cfg.from_date,
    to_date=cfg.to_date,
    symbols=symbols,
    trades=len(pnls),
    win_rate=(len([x for x in pnls if x > 0]) / len(pnls)) if pnls else 0.0,
    total_pnl=realized,
    max_drawdown_abs=_max_drawdown_abs(equity_curve),
    max_drawdown_pct=_max_drawdown_pct(equity_curve),
    cagr_pct=_cagr(equity_curve, cfg.initial_capital, final_cap),
    sharpe_proxy=((sma(returns) / std(returns)) * (252**0.5)) if returns and std(returns) > 0 else 0.0,
  )


def _build_candidates(day: str, bars_by_symbol: dict[str, list[dict[str, float | str]]], cfg: BacktestConfig) -> list[dict[str, float | str]]:
  out: list[dict[str, float | str]] = []
  for symbol, bars in bars_by_symbol.items():
    idx = next((i for i, b in enumerate(bars) if str(b["time"]).startswith(day)), -1)
    if idx < 61:
      continue
    s = bars[: idx + 1]
    if len(s) < 80:
      continue
    closes = [float(x["close"]) for x in s]
    highs = [float(x["high"]) for x in s]
    lows = [float(x["low"]) for x in s]
    volumes = [float(x["volume"]) for x in s]
    close = closes[-1]
    ema20 = ema(closes, 20)
    ema50 = ema(closes, 50)
    rsi14 = rsi(closes, 14)
    if not (close > ema20 > ema50 and rsi14 >= cfg.min_rsi):
      continue
    high20 = max(highs[-20:])
    breakout = high20 * (1 + cfg.breakout_buffer_pct)
    if close < breakout:
      continue
    atr14 = atr(highs, lows, closes, 14)
    stop = close - atr14 * cfg.atr_stop_multiple
    risk_per_share = max(close - stop, 0.01)
    qty = max(1, int((cfg.initial_capital * cfg.risk_per_trade) / risk_per_share))
    target = close + (close - stop) * 2
    rs = pct_change(closes[-61], close)
    vol_ratio = sma(volumes[-20:]) / max(1, sma(volumes[-50:]))
    out.append(
      {
        "symbol": symbol,
        "entry": close,
        "stop": stop,
        "target": target,
        "qty": qty,
        "rank": rs * vol_ratio,
      }
    )
  out.sort(key=lambda x: float(x["rank"]), reverse=True)
  return out


def _load_bars(
  provider: KiteHistoricalProvider,
  symbols: list[str],
  history_from: str,
  from_date: str,
  to_date: str,
) -> dict[str, list[dict[str, float | str]]]:
  instruments = provider.get_instruments("NSE")
  out: dict[str, list[dict[str, float | str]]] = {}
  for symbol in symbols:
    inst = instruments.get(symbol)
    if not inst:
      continue
    bars = provider.get_historical_day_bars(inst.instrument_token, history_from, to_date)
    normalized = [
      {
        "time": b.time,
        "open": b.open,
        "high": b.high,
        "low": b.low,
        "close": b.close,
        "volume": b.volume,
      }
      for b in bars
      if b.close and b.volume
    ]
    if any(from_date <= str(x["time"])[:10] <= to_date for x in normalized):
      out[symbol] = sorted(normalized, key=lambda x: str(x["time"]))
  return out


def _offset_from(yyyy_mm_dd: str, days_back: int) -> str:
  d = datetime.fromisoformat(yyyy_mm_dd).replace(tzinfo=UTC) - timedelta(days=days_back)
  return d.date().isoformat()


def _slippage(price: float, side: str, bps: float) -> float:
  factor = bps / 10_000
  return price * (1 + factor) if side == "BUY" else price * (1 - factor)


def _max_drawdown_abs(series: list[float]) -> float:
  peak = float("-inf")
  out = 0.0
  for v in series:
    peak = max(peak, v)
    out = max(out, peak - v)
  return out


def _max_drawdown_pct(series: list[float]) -> float:
  peak = float("-inf")
  out = 0.0
  for v in series:
    peak = max(peak, v)
    if peak > 0:
      out = max(out, (peak - v) / peak)
  return out


def _cagr(equity_curve: list[float], initial: float, final: float) -> float:
  if len(equity_curve) < 2 or initial <= 0 or final <= 0:
    return 0.0
  years = max(len(equity_curve) / 252, 1 / 252)
  return ((final / initial) ** (1 / years) - 1) * 100
